20240328-queries-with-usersets.md

```diff
@@ -0,0 +1,176 @@
+| -                                   | Check(document:1#difference_1@document:1#a)               | `{allowed=false}` | `{allowed=false}` (no change) |
+| -                                   | Check(document:1#difference_1@document:1#a)               | `{allowed=false}` | `{allowed=false}` (no change) |
+| document:1#c@group:marketing#member | Check(document:1#difference_2@group:marketing#member)     | `{allowed=true}`  | `{allowed=true}`  (no change) |
+
+
+https://play.fga.dev/stores/create/?id=01HT3QDCENG2D0A3KS0CNQSTAJ.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Check API involves solving a parent problem that may contain subproblems. The implementation of this proposal will be that when solving the parent problem or each subproblem, if the problem is of the form `objectType:objectId # relation @ objectType:objectId#relation`, then we will immediately return `{allowed=true}`.
+
+The same idea will apply to ListObjects and ListUsers APIs.
+
+# Migration
+[migration]: #migration
+
+This proposal doesn't entail any migration of data. However, it will require a thorough review by application developers and authorization model authors to ensure that their applications are making the right queries to OpenFGA. Some developers making Check API calls with a userset in the `user` field may want to change this to be a specific user, as the application may start granting more permissions than desired.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Confusion for application developers. So far, developers have largely assumed that a Check returns `{allowed: true}` if and only if there is a tuple or tuples in the system saying so. With this proposal we are introducing the idea of "implicit" tuples. 
+
+# Alternatives
+[alternatives]: #alternatives
+
+N/A
+
+# Prior Art
+[prior-art]: #prior-art
+
+N/A
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Given that the userset `document:1#viewer` is defined as the set that has the `viewer` relation with `document:1`, should we block application developers from writing the tuple `document:1#viewer@document:1#viewer` since the Write will become irrelevant? And if yes, what error should we return?
+- How does this affect the [Read API](https://openfga.dev/api/service#/Relationship%20Tuples/Read)? The documentation says 
+    > The Read API will return the tuples for a certain store that match a query filter specified in the body of the request. [...] it only returns relationship tuples that are stored in the system and satisfy the query.
+
+    If a developer calls `Read(user=document:1#viewer, relation:viewer, object=document:1)` today, unless they previously wrote the tuple `document:1 # viewer @ document:1#viewer`, this query returns an empty result, even though the tuple will be assumed to exist.
+- How does this affect the [Expand API](https://openfga.dev/api/service#/Relationship%20Queries/Expand)? 
```

Comment:
This is one that worries me. We cannot return this tuple in Expand, as most devs recursively calling Expand will follow it again and end up in an infinite recursion.

On the other hand it means drawing the graph for the users is going to be harder.

And if we draw this on every relation, we are needlessly complicating the graph UI.

Reply:
https://github.com/openfga/openfga/pull/1522